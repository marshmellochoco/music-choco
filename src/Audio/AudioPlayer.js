import {
    useState,
    useEffect
} from "react";

export default function AudioPlayer() {
    const [duration, setDuration] = useState()
    const [curTime, setCurTime] = useState()
    const [volume, setVolume] = useState()
    const [clickedTime, setClickedTime] = useState()
    const [playing, setPlaying] = useState(false)

    useEffect(() => {
        const audio = document.getElementById('audio-player')

        const setAudioData = () => {
            setDuration(audio.duration)
            setCurTime(audio.currentTime)
            setVolume(audio.volume)
        }

        const setAudioTime = () => setCurTime(audio.currentTime)
        const setAudioVolume = () => setVolume(audio.volume)

        audio.addEventListener("loadeddata", setAudioData)
        audio.addEventListener("timeupdate", setAudioTime)
        audio.addEventListener("volumechange", setAudioVolume)

        playing ? audio.play() : audio.pause()

        if (clickedTime && clickedTime !== curTime) {
            audio.currentTime = clickedTime
            setClickedTime(null);
        }

        if (volume && volume !== audio.volume)
            audio.volume = volume; 

        return () => {
            audio.removeEventListener("loadeddata", setAudioData)
            audio.removeEventListener("timeupdate", setAudioTime)
            audio.removeEventListener("volumechange", setAudioVolume)
        }
    }, [playing, clickedTime, curTime, volume])

    return {
        curTime,
        duration,
        playing,
        volume,
        setPlaying,
        setClickedTime,
        setVolume
    }

}
