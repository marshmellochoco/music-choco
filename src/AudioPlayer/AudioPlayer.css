.player{
    font-size: 0.8em;
    width: 98%;
    padding: 10px 1%;
    background-color: #303030;
    color: #d0d0d0;
    overflow: hidden;
    position: absolute;
    bottom: 0;
}

.player-container {
    display: grid;
    grid-template-columns: 1fr 2fr 1fr;
    grid-template-rows: 2fr 1fr;
}

.player-title{
    margin: auto 0;
    grid-row: 1/3;
    text-align: left;
    display: grid;
    grid-template-columns: 64px 1fr;
    grid-template-rows: 1fr 1fr;
}

.player-title .icon{
    grid-row: 1/3;
    width: 64px;
    height: 64px;
}

.player-title .icon img{
    max-width: 64px;
    max-height: 64px;
    background-color: white;
}

.player-title .title{
    margin: 0 10px;
    display: flex;
    align-items: flex-end;
}

.player-title .artist{
    margin: 0 10px;
    display: flex;
    align-items: flex-start;
}


.player-timeline{
    margin: auto 0;
    text-align: center;
    display: grid;
    grid-template-columns: 1fr 5fr 1fr;
}

.player-timeline .progress{
    overflow: hidden;
    margin: auto 0;
    background-color: grey;
    height: 0.3em;
    border-radius: 2px;
    cursor: pointer;
}

.player-timeline .player-time{
    padding: 0 10%;
    font-size: 0.9em;
}

/* 
.player-timeline input[type=range]{
    overflow: hidden;
    width: 70%;
    -webkit-appearance: none;
    background-color: #303030;
}

input[type=range]::-webkit-slider-runnable-track{
    border: none;
    height: 0.35em;
    background: grey;
    border-radius: 10px;
    cursor: pointer;
}

input[type=range]::-webkit-slider-thumb{
    -webkit-appearance: none;
    height: 0.5em;
    width: 0.5em;
    border-radius: 50%;
    background: white;
    margin-top: -1px;
    cursor: pointer;
    box-shadow: -80px 0 0 80px green;
}

input[type=range]::-webkit-slider-thumb:hover{
    height: 0.8em;
    width: 0.8em;
    border-radius: 50%;
    background: white;
    margin-top: -3px;
}

input[type=range]:focus {
    outline: none;
} */

.player-control{
    text-align: center;
}

.player-control .control-button{
    height: 100%;
    margin: auto;
    width: 1.8em;
    padding: 0px 12px;
}

.player-control .play{
    width: 2.5em;
}

.player-control .control-button:hover{
    color: grey;
}

.player-volume{
    display: flex;
    margin-left: auto;
    width: 30%;
    grid-row: 1/3;
    grid-column: 3;
}

@media only screen and (max-width: 1024px){
    .player-volume input{
        width: 50%;
    }
}

@media only screen and (max-width: 862px){
    input[type=range]{
        width: 60%;
    }
}

@media only screen and (max-width: 480px){
    .player-time{
        display: none;
    }
    .player-timeline input[type=range]{
        width: 100%;
    }
    .player-control .control-button{
        font-size: 1em
    }
}

.slider{
    overflow: hidden;
    margin: auto 0;
    background-color: grey;
    height: 0.3em;
    border-radius: 2px;
    cursor: pointer;
}

.slider input[type=range]{
    cursor: pointer;
    opacity: 0;
    margin: auto 0;
    float: right;
}