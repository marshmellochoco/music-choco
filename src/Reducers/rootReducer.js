import { queueReducer } from "../Reducers/queueReducer";
import { songDataReducer } from "../Reducers/songDataReducer";
import { playerReducer } from "../Reducers/playerReducer";
import { authReducer } from "../Reducers/authReducer";

import { combineReducers } from "redux";

export const rootReducer = combineReducers({
    queueReducer,
    songDataReducer,
    playerReducer,
    authReducer,
});
