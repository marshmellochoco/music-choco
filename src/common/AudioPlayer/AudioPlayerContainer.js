import { AudioPlayerComponent } from "./AudioPlayerComponent";
import { createRef, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { setPlayingSong } from "../../store/actions/songDataAction";

export const AudioPlayerContainer = () => {
    // initialization
    const apiUrl = process.env.REACT_APP_API_URL;
    const songData = useSelector((state) => state.songDataReducer.songData);
    const queue = useSelector((state) => state.queueReducer.queue);
    const playing = useSelector((state) => state.playerReducer.playing);
    const volume = useSelector((state) => state.playerReducer.volume);
    const isLoop = useSelector((state) => state.playerReducer.loop);
    const isRandom = useSelector((state) => state.playerReducer.random);
    const [currentTime, setCurrentTime] = useState(0);
    const [lastVolume, setLastVolume] = useState(volume);
    const dispatch = useDispatch();
    const ref = createRef();

    // functions
    // player functions
    const onProgress = (playedSeconds) => {
        // set the current time as the number of seconds played on the song
        if (Math.floor(playedSeconds) <= songData.duration) {
            setCurrentTime(Math.ceil(playedSeconds));
        }
    };

    const onEnded = () => {
        // play the next song when the current song has ended, reset the played time
        dispatch({ type: "SET_PLAYING", playing: false });
        setCurrentTime(0);
        nextSong();
        dispatch({ type: "SET_PLAYING", playing: true });
    };

    // control functions
    const setClickedTime = (percent) => {
        // seek to the partition of the song according the the percentage of timeline
        let second = (percent / 100) * songData.duration;
        ref.current.seekTo(second, "second");
    };

    const changeVolume = (vol) => {
        // change the volume, if it is unmuted, set volume to the previous volume
        setLastVolume(volume === 0 ? lastVolume : volume);
        dispatch({ type: "SET_VOLUME", volume: vol });
        console.log(vol);
    };

    const prevSong = () => {
        // play the previous song, if no previous song, restart the song
        // if is random, use the random queue list else use the normal queue list
        if (isRandom) {
            if (randomQueue.indexOf(songData.songId) === 0) {
                setClickedTime(0);
            } else {
                dispatch(
                    setPlayingSong(
                        randomQueue[randomQueue.indexOf(songData.songId) - 1]
                    )
                );
            }
        } else {
            if (queue.indexOf(songData.songId) === 0) {
                setClickedTime(0);
            } else {
                dispatch(
                    setPlayingSong(queue[queue.indexOf(songData.songId) - 1])
                );
            }
        }
    };

    const nextSong = () => {
        // play the next song, if is the last song in the list, play nothing
        // if is random, use the random queue list else use the normal queue list
        if (!songData.songId) return;

        const next = (q) => {
            if (q.indexOf(songData.songId) === q.length - 1) {
                dispatch(setPlayingSong(isLoop ? q[0] : ""));
            } else {
                dispatch(setPlayingSong(q[q.indexOf(songData.songId) + 1]));
            }
        };

        next(isRandom ? randomQueue : queue);
    };

    return (
        <AudioPlayerComponent
            songUrl={
                songData.songId ? `${apiUrl}/song/play/${songData.songId}` : ""
            }
            albumUrl={
                songData && songData.songId
                    ? `${apiUrl}/album/ico/${songData.albumId}`
                    : ""
            }
            playing={playing}
            volume={volume}
            lastVolume={lastVolume}
            songData={songData}
            currentTime={currentTime}
            setCurrentTime={setCurrentTime}
            onProgress={onProgress}
            onEnded={onEnded}
            nextSong={nextSong}
            prevSong={prevSong}
            changeVolume={changeVolume}
            setClickedTime={setClickedTime}
        />
    );
};
