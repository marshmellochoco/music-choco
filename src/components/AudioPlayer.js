import { useState, useLayoutEffect } from "react";
import { useDispatch } from "react-redux";
import { useSelector } from "react-redux";
import AudioPlayerComponent from "./AudioPlayerComponent";
import MediaSession from "./MediaSession";
import { setPlaying, setPlayingTrack } from "../store/player/playerAction";

const AudioPlayer = ({ openQueue, setOpenQueue, showPlayer }) => {
    const seekTime = 5;
    var ref = "";
    const dispatch = useDispatch();
    const { playingTrack, playing, loop } = useSelector(
        (state) => state.playerReducer
    );
    const queue = useSelector((state) => state.queueReducer);
    const [time, setTime] = useState(0);
    const [media, setMedia] = useState(playingTrack);

    useLayoutEffect(() => {
        if (playing) ref.current.play();
        else ref.current.pause();
    }, [playing, ref]);

    const setRef = (input) => {
        ref = input;
    };

    const onPlay = () => {
        if (!ref.current) return;
        ref.current.play();
    };

    const onPause = () => {
        if (!ref.current) return;
        ref.current.pause();
    };

    const onPreviousTrack = () => {
        const index = queue.indexOf(playingTrack) - 1;
        if (index >= 0) {
            dispatch(setPlayingTrack(queue[index]));
            dispatch(setPlaying(true));
        } else {
            ref.current.currentTime = 0;
            setTime(0);
        }
    };

    const onNextTrack = () => {
        const index = queue.indexOf(playingTrack) + 1;
        if (index >= queue.length) {
            if (!loop) {
                dispatch(setPlaying(false));
                ref.current.currentTime = 0;
                setTime(100);
            } else {
                dispatch(setPlayingTrack(queue[0]));
                dispatch(setPlaying(true));
            }
        } else {
            dispatch(setPlayingTrack(queue[index]));
            dispatch(setPlaying(true));
        }
    };

    const seekByTime = (t) => {
        if (!ref.current) return;
        ref.current.currentTime += t;
    };

    const seekPrecent = (progress) => {
        if (!ref.current) return;
        ref.current.currentTime =
            (Math.ceil(ref.current.duration) - 1) * progress;
    };

    useLayoutEffect(() => {
        setMedia({
            ...playingTrack,
        });
    }, [playingTrack]);

    return (
        <>
            {media.artists && (
                <MediaSession
                    title={media.title}
                    artist={media.artists.map((a) => a.name).join(", ")}
                    album={media.album.name}
                    artwork={[
                        {
                            src: media.album.image,
                        },
                    ]}
                    onPlay={onPlay}
                    onPause={onPause}
                    onNextTrack={onNextTrack}
                    onPreviousTrack={onPreviousTrack}
                    onSeekForward={() => seekByTime(seekTime)}
                    onSeekBackward={() => seekByTime(-seekTime)}
                />
            )}

            <AudioPlayerComponent
                openQueue={openQueue}
                setOpenQueue={setOpenQueue}
                time={time}
                setRef={setRef}
                onPlay={onPlay}
                onPause={onPause}
                onPreviousTrack={onPreviousTrack}
                onNextTrack={onNextTrack}
                seekPrecent={seekPrecent}
                showPlayer={showPlayer}
            />
        </>
    );
};

export default AudioPlayer;
