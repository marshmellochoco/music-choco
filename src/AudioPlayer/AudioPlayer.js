import AudioTimeline from './PlayerTimeline';
import VolumeSlider from './VolumeSlider';
import AudioControl from './PlayerControl';
import './AudioPlayer.css';
import AudioPlaying from './NowPlaying';
import { useState, useEffect } from "react";

export default function Audio() {
    const {
        curTime,
        duration,
        playing,
        volume,
        setPlaying,
        setClickedTime,
        setVolume
    } = AudioPlayer();

    const song1 = {
        icon: 'https://i.scdn.co/image/ab67616d00004851bd5899a4711f4604447166c3',
        title: 'TITLE',
        artist: 'ARTIST',
        song: 'https://www.soundhelix.com/examples/mp3/SoundHelix-Song-1.mp3'
    };


    function handleKeyDown(key){
        switch(key){
            case 'ArrowLeft':
                setClickedTime(curTime - 5);
                break;
            case 'ArrowRight':
                setClickedTime(curTime + 5);
                break;
            case 'ArrowDown':
                setVolume(volume > 0.1 ? volume - 0.1 : 0);
                break;
            case 'ArrowUp':
                setVolume(volume < 0.9 ? volume + 0.1 : 1);
                break;
            case ' ':
                setPlaying(!playing);
                break;
            default:
                return
        }
    }

    return (
        <div className='player' onKeyDown={(e) => handleKeyDown(e.key)}>
            <audio id='audio-player' src={song1.song}/>
            <div className='player-container'>
                <AudioPlaying icon={song1.icon} title={song1.title} artist={song1.artist}/>
                <AudioControl playing={playing} setPlaying={setPlaying}/>
                <VolumeSlider volume={volume} setVolume={(volume) => setVolume(volume)}/>
                <AudioTimeline duration={duration} curTime={curTime}/>
            </div>
        </div>
    )
}

function AudioPlayer() {
    const [duration, setDuration] = useState()
    const [curTime, setCurTime] = useState()
    const [volume, setVolume] = useState()
    const [clickedTime, setClickedTime] = useState()
    const [playing, setPlaying] = useState(false)

    useEffect(() => {
        const audio = document.getElementById('audio-player')

        const setAudioData = () => {
            setDuration(audio.duration)
            setCurTime(audio.currentTime)
            setVolume(audio.volume)
        }

        const setAudioTime = () => setCurTime(audio.currentTime)
        const setAudioVolume = () => setVolume(audio.volume)

        audio.addEventListener("loadeddata", setAudioData)
        audio.addEventListener("timeupdate", setAudioTime)
        audio.addEventListener("volumechange", setAudioVolume)

        playing ? audio.play() : audio.pause()

        if (clickedTime && clickedTime !== curTime) {
            audio.currentTime = clickedTime
            setClickedTime(null);
        }

        if (volume && volume !== audio.volume)
            audio.volume = volume; 

        return () => {
            audio.removeEventListener("loadeddata", setAudioData)
            audio.removeEventListener("timeupdate", setAudioTime)
            audio.removeEventListener("volumechange", setAudioVolume)
        }
    }, [playing, clickedTime, curTime, volume])

    return {
        curTime,
        duration,
        playing,
        volume,
        setPlaying,
        setClickedTime,
        setVolume
    }
}