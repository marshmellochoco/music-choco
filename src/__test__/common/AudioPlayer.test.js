import { configure, mount } from "enzyme";
import Adapter from "@wojtekmaj/enzyme-adapter-react-17";
import toJson from "enzyme-to-json";
import { Provider } from "react-redux";
import { createStore } from "redux";

import { rootReducer } from "../../store/rootReducer";
import { AudioPlayerContainer } from "../../common/AudioPlayer/AudioPlayerContainer";

configure({ adapter: new Adapter() });
const store = createStore(rootReducer);

describe("AudioPlayer", () => {
    it("should render the player", () => {
        const wrapper = mount(
            <Provider store={store}>
                <AudioPlayerContainer />
            </Provider>
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });
});
