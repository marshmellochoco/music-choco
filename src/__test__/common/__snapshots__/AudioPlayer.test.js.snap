// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AudioPlayer should render the player 1`] = `
<Provider
  store={
    Object {
      "@@observable": [Function],
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <AudioPlayerContainer>
    <audio
      onCanPlay={[Function]}
      onEnded={[Function]}
      onTimeUpdate={[Function]}
      src=""
    />
    <AudioPlayerComponent>
      <EmotionCssPropInternal
        __EMOTION_LABEL_PLEASE_DO_NOT_USE__="AudioPlayerComponent"
        __EMOTION_TYPE_PLEASE_DO_NOT_USE__="div"
        css={
          Object {
            "map": undefined,
            "name": "1sbj9sg",
            "next": undefined,
            "styles": "
		font-size: 0.8em;
		width: 98%;
		padding: 10px 1%;
		background-color: #303030;
		color: #d0d0d0;
		overflow: hidden;
		position: absolute;
		bottom: 0;

		&:focus {
			outline: none;
		}
	",
            "toString": [Function],
          }
        }
      >
        <div
          className="css-317edk-AudioPlayerComponent"
        >
          <EmotionCssPropInternal
            __EMOTION_LABEL_PLEASE_DO_NOT_USE__="AudioPlayerComponent"
            __EMOTION_TYPE_PLEASE_DO_NOT_USE__="div"
            css={
              Object {
                "map": undefined,
                "name": "5bg9bn",
                "next": undefined,
                "styles": "
		display: grid;
		grid-template-columns: minmax(2rem, 1fr) minmax(4rem, 2fr) minmax(2rem, 0.8fr);
		grid-template-rows: 1fr 1fr;
	",
                "toString": [Function],
              }
            }
          >
            <div
              className="css-rb6vzw-AudioPlayerComponent"
            >
              <PlayerDetailComponent
                albumUrl=""
                handleImageError={[Function]}
                songData={
                  Object {
                    "albumId": "",
                    "albumname": "",
                    "artist": "",
                    "duration": 0,
                    "songId": "",
                    "title": "",
                  }
                }
              >
                <EmotionCssPropInternal
                  __EMOTION_LABEL_PLEASE_DO_NOT_USE__="PlayerDetailComponent"
                  __EMOTION_TYPE_PLEASE_DO_NOT_USE__="div"
                  css={
                    Object {
                      "map": undefined,
                      "name": "ms89kb",
                      "next": undefined,
                      "styles": "
		margin: auto 0;
		grid-row: 1/3;
		text-align: left;
		display: grid;
		grid-template-columns: 64px 1fr;
		grid-template-rows: 1fr 1fr;
		@media (max-width: 768px) {
			display: flex;
			flex-direction: column;
		}
	",
                      "toString": [Function],
                    }
                  }
                >
                  <div
                    className="css-1sq3i5d-PlayerDetailComponent"
                  >
                    <EmotionCssPropInternal
                      __EMOTION_LABEL_PLEASE_DO_NOT_USE__="PlayerDetailComponent"
                      __EMOTION_TYPE_PLEASE_DO_NOT_USE__="div"
                      css={
                        Object {
                          "map": undefined,
                          "name": "ddduf2",
                          "next": undefined,
                          "styles": "
		grid-row: 1/3;
		width: 64px;
		height: 64px;

		& > img {
			max-width: 64px;
			max-height: 64px;
			background-color: white;
		}

		@media (max-width: 768px) {
			display: none;
		}
	",
                          "toString": [Function],
                        }
                      }
                    >
                      <div
                        className="css-1ym2h18-PlayerDetailComponent"
                      >
                        <img
                          onError={[Function]}
                          src=""
                        />
                      </div>
                    </EmotionCssPropInternal>
                    <EmotionCssPropInternal
                      __EMOTION_LABEL_PLEASE_DO_NOT_USE__="PlayerDetailComponent"
                      __EMOTION_TYPE_PLEASE_DO_NOT_USE__="div"
                      css={
                        Object {
                          "map": undefined,
                          "name": "70z0j6",
                          "next": undefined,
                          "styles": "
		font-size: 1.2em;
		margin: 0 10px;
		display: flex;
		align-items: flex-end;
		white-space: nowrap;
		overflow: hidden !important;
		text-overflow: ellipsis;
	",
                          "toString": [Function],
                        }
                      }
                    >
                      <div
                        className="css-677yt3-PlayerDetailComponent"
                      >
                        ---
                      </div>
                    </EmotionCssPropInternal>
                    <EmotionCssPropInternal
                      __EMOTION_LABEL_PLEASE_DO_NOT_USE__="PlayerDetailComponent"
                      __EMOTION_TYPE_PLEASE_DO_NOT_USE__="div"
                      css={
                        Object {
                          "map": undefined,
                          "name": "zojdeh",
                          "next": undefined,
                          "styles": "
		font-size: 1em;
		margin: 0 10px;
		display: flex;
		align-items: flex-start;
		white-space: nowrap;
		overflow: hidden !important;
		text-overflow: ellipsis;
	",
                          "toString": [Function],
                        }
                      }
                    >
                      <div
                        className="css-10fvclc-PlayerDetailComponent"
                      >
                        ---
                      </div>
                    </EmotionCssPropInternal>
                  </div>
                </EmotionCssPropInternal>
              </PlayerDetailComponent>
              <PlayerControlComponent
                nextSong={[Function]}
                playPause={[Function]}
                playing={false}
                prevSong={[Function]}
              >
                <EmotionCssPropInternal
                  __EMOTION_LABEL_PLEASE_DO_NOT_USE__="PlayerControlComponent"
                  __EMOTION_TYPE_PLEASE_DO_NOT_USE__="div"
                  css={
                    Object {
                      "map": undefined,
                      "name": "1czjaxl",
                      "next": undefined,
                      "styles": "
		text-align: center;

		& > svg {
			height: 100%;
			margin: auto;
			width: 1.8em;
			margin: 0px 12px;
			cursor: pointer;

			&:hover {
				opacity: 0.7;
			}
		}

		& > .play {
			width: 2.5em;
		}
	",
                      "toString": [Function],
                    }
                  }
                >
                  <div
                    className="css-ezed4n-PlayerControlComponent"
                  >
                    <Icon
                      color="currentColor"
                      horizontal={false}
                      onClick={[Function]}
                      path="M6,18V6H8V18H6M9.5,12L18,6V18L9.5,12Z"
                      rotate={0}
                      size={null}
                      spin={false}
                      vertical={false}
                    >
                      <svg
                        onClick={[Function]}
                        role="presentation"
                        style={Object {}}
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M6,18V6H8V18H6M9.5,12L18,6V18L9.5,12Z"
                          style={
                            Object {
                              "fill": "currentColor",
                            }
                          }
                        />
                      </svg>
                    </Icon>
                    <Icon
                      className="play"
                      color="currentColor"
                      horizontal={false}
                      onClick={[Function]}
                      path="M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M10,16.5L16,12L10,7.5V16.5Z"
                      rotate={0}
                      size={null}
                      spin={false}
                      vertical={false}
                    >
                      <svg
                        className="play"
                        onClick={[Function]}
                        role="presentation"
                        style={Object {}}
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M10,16.5L16,12L10,7.5V16.5Z"
                          style={
                            Object {
                              "fill": "currentColor",
                            }
                          }
                        />
                      </svg>
                    </Icon>
                    <Icon
                      color="currentColor"
                      horizontal={false}
                      onClick={[Function]}
                      path="M16,18H18V6H16M6,18L14.5,12L6,6V18Z"
                      rotate={0}
                      size={null}
                      spin={false}
                      vertical={false}
                    >
                      <svg
                        onClick={[Function]}
                        role="presentation"
                        style={Object {}}
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M16,18H18V6H16M6,18L14.5,12L6,6V18Z"
                          style={
                            Object {
                              "fill": "currentColor",
                            }
                          }
                        />
                      </svg>
                    </Icon>
                  </div>
                </EmotionCssPropInternal>
              </PlayerControlComponent>
              <PlayerToolComponent
                changeVolume={[Function]}
                lastVolume={1}
                toggleQueue={[Function]}
                volume={1}
              >
                <EmotionCssPropInternal
                  __EMOTION_LABEL_PLEASE_DO_NOT_USE__="PlayerToolComponent"
                  __EMOTION_TYPE_PLEASE_DO_NOT_USE__="div"
                  css={
                    Object {
                      "map": undefined,
                      "name": "c56ff1",
                      "next": undefined,
                      "styles": "
		display: flex;
		justify-content: flex-end;
		grid-row: 1/3;
		grid-column: 3;

		& svg {
			padding: 4px;
			width: 1.8em;
			cursor: pointer;
			margin: 0 10px;

			&:hover {
				opacity: 0.7;
			}
		}

		& > div {
			display: flex;
			& > div {
				@media (max-width: 768px) {
					display: none;
				}
			}
		}
	",
                      "toString": [Function],
                    }
                  }
                >
                  <div
                    className="css-r0qo89-PlayerToolComponent"
                  >
                    <div>
                      <Icon
                        color="currentColor"
                        horizontal={false}
                        onClick={[Function]}
                        path="M3,3H21V7H3V3M4,8H20V21H4V8M9.5,11A0.5,0.5 0 0,0 9,11.5V13H15V11.5A0.5,0.5 0 0,0 14.5,11H9.5Z"
                        rotate={0}
                        size={null}
                        spin={false}
                        vertical={false}
                      >
                        <svg
                          onClick={[Function]}
                          role="presentation"
                          style={Object {}}
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M3,3H21V7H3V3M4,8H20V21H4V8M9.5,11A0.5,0.5 0 0,0 9,11.5V13H15V11.5A0.5,0.5 0 0,0 14.5,11H9.5Z"
                            style={
                              Object {
                                "fill": "currentColor",
                              }
                            }
                          />
                        </svg>
                      </Icon>
                    </div>
                    <div>
                      <Icon
                        color="currentColor"
                        horizontal={false}
                        onClick={[Function]}
                        path="M14,3.23V5.29C16.89,6.15 19,8.83 19,12C19,15.17 16.89,17.84 14,18.7V20.77C18,19.86 21,16.28 21,12C21,7.72 18,4.14 14,3.23M16.5,12C16.5,10.23 15.5,8.71 14,7.97V16C15.5,15.29 16.5,13.76 16.5,12M3,9V15H7L12,20V4L7,9H3Z"
                        rotate={0}
                        size={null}
                        spin={false}
                        vertical={false}
                      >
                        <svg
                          onClick={[Function]}
                          role="presentation"
                          style={Object {}}
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M14,3.23V5.29C16.89,6.15 19,8.83 19,12C19,15.17 16.89,17.84 14,18.7V20.77C18,19.86 21,16.28 21,12C21,7.72 18,4.14 14,3.23M16.5,12C16.5,10.23 15.5,8.71 14,7.97V16C15.5,15.29 16.5,13.76 16.5,12M3,9V15H7L12,20V4L7,9H3Z"
                            style={
                              Object {
                                "fill": "currentColor",
                              }
                            }
                          />
                        </svg>
                      </Icon>
                      <Slider
                        onChange={[Function]}
                        transform={0}
                      >
                        <EmotionCssPropInternal
                          __EMOTION_LABEL_PLEASE_DO_NOT_USE__="Slider"
                          __EMOTION_TYPE_PLEASE_DO_NOT_USE__="div"
                          css={
                            Object {
                              "map": undefined,
                              "name": "1s3qkif",
                              "next": undefined,
                              "styles": "
		width: 96%;
		margin: auto 2%;
		background-color: grey;
		height: 0.3em;
		border-radius: 2px;
		overflow: hidden;
	",
                              "toString": [Function],
                            }
                          }
                        >
                          <div
                            className="css-xowwt8-Slider"
                          >
                            <EmotionCssPropInternal
                              __EMOTION_LABEL_PLEASE_DO_NOT_USE__="Slider"
                              __EMOTION_TYPE_PLEASE_DO_NOT_USE__="div"
                              css={
                                Object {
                                  "map": undefined,
                                  "name": "zvp377",
                                  "next": undefined,
                                  "styles": "
		width: 100%;
		height: 100%;
		background-color: green;
	",
                                  "toString": [Function],
                                }
                              }
                              style={
                                Object {
                                  "transform": "translateX(0%)",
                                }
                              }
                            >
                              <div
                                className="css-y9lou1-Slider"
                                style={
                                  Object {
                                    "transform": "translateX(0%)",
                                  }
                                }
                              />
                            </EmotionCssPropInternal>
                            <EmotionCssPropInternal
                              __EMOTION_LABEL_PLEASE_DO_NOT_USE__="Slider"
                              __EMOTION_TYPE_PLEASE_DO_NOT_USE__="input"
                              css={
                                Object {
                                  "map": undefined,
                                  "name": "119gc6m",
                                  "next": undefined,
                                  "styles": "
		cursor: pointer;
		width: 100%;
		height: 100%;
		opacity: 0;
		margin: auto 0;
		float: right;
		transform: translateY(-100%);
	",
                                  "toString": [Function],
                                }
                              }
                              onChange={[Function]}
                              type="range"
                            >
                              <input
                                className="css-6plv7s-Slider"
                                onChange={[Function]}
                                type="range"
                              />
                            </EmotionCssPropInternal>
                          </div>
                        </EmotionCssPropInternal>
                      </Slider>
                    </div>
                  </div>
                </EmotionCssPropInternal>
              </PlayerToolComponent>
              <PlayerTimelineComponent
                currentTime={0}
                duration={0}
                setClickedTime={[Function]}
              >
                <EmotionCssPropInternal
                  __EMOTION_LABEL_PLEASE_DO_NOT_USE__="PlayerTimelineComponent"
                  __EMOTION_TYPE_PLEASE_DO_NOT_USE__="div"
                  css={
                    Object {
                      "map": undefined,
                      "name": "16eldsh",
                      "next": undefined,
                      "styles": "
		margin: auto 0;
		text-align: center;
		display: grid;
		grid-template-columns: 4rem 10fr 4rem;
		font-size: 0.9em;
	",
                      "toString": [Function],
                    }
                  }
                >
                  <div
                    className="css-4nwngi-PlayerTimelineComponent"
                  >
                    <div>
                      0 
                    0
                : 0 0
                    </div>
                    <Slider
                      onChange={[Function]}
                      transform={NaN}
                    >
                      <EmotionCssPropInternal
                        __EMOTION_LABEL_PLEASE_DO_NOT_USE__="Slider"
                        __EMOTION_TYPE_PLEASE_DO_NOT_USE__="div"
                        css={
                          Object {
                            "map": undefined,
                            "name": "1s3qkif",
                            "next": undefined,
                            "styles": "
		width: 96%;
		margin: auto 2%;
		background-color: grey;
		height: 0.3em;
		border-radius: 2px;
		overflow: hidden;
	",
                            "toString": [Function],
                          }
                        }
                      >
                        <div
                          className="css-xowwt8-Slider"
                        >
                          <EmotionCssPropInternal
                            __EMOTION_LABEL_PLEASE_DO_NOT_USE__="Slider"
                            __EMOTION_TYPE_PLEASE_DO_NOT_USE__="div"
                            css={
                              Object {
                                "map": undefined,
                                "name": "zvp377",
                                "next": undefined,
                                "styles": "
		width: 100%;
		height: 100%;
		background-color: green;
	",
                                "toString": [Function],
                              }
                            }
                            style={
                              Object {
                                "transform": "translateX(NaN%)",
                              }
                            }
                          >
                            <div
                              className="css-y9lou1-Slider"
                              style={
                                Object {
                                  "transform": "translateX(NaN%)",
                                }
                              }
                            />
                          </EmotionCssPropInternal>
                          <EmotionCssPropInternal
                            __EMOTION_LABEL_PLEASE_DO_NOT_USE__="Slider"
                            __EMOTION_TYPE_PLEASE_DO_NOT_USE__="input"
                            css={
                              Object {
                                "map": undefined,
                                "name": "119gc6m",
                                "next": undefined,
                                "styles": "
		cursor: pointer;
		width: 100%;
		height: 100%;
		opacity: 0;
		margin: auto 0;
		float: right;
		transform: translateY(-100%);
	",
                                "toString": [Function],
                              }
                            }
                            onChange={[Function]}
                            type="range"
                          >
                            <input
                              className="css-6plv7s-Slider"
                              onChange={[Function]}
                              type="range"
                            />
                          </EmotionCssPropInternal>
                        </div>
                      </EmotionCssPropInternal>
                    </Slider>
                    <div>
                      0 
                    0
                : 0 0
                    </div>
                  </div>
                </EmotionCssPropInternal>
              </PlayerTimelineComponent>
            </div>
          </EmotionCssPropInternal>
        </div>
      </EmotionCssPropInternal>
    </AudioPlayerComponent>
  </AudioPlayerContainer>
</Provider>
`;
